### DISCUSSION TITLE
Frontend vs. API-Route problem (transfer values between them)

### DISCUSSION THREAD

**Discussion by gregorsart** [neuefische/questions#763] [Web] [answered]
We have three questions so far, regarding barking, trainability and energy. We want to fetch from the api via API-Routes with these "key-words". For example `https://api.api-ninjas.com/v1/dogs?barking=3`.

How to transfer a string-value (for example barking) and its value (for example 3)  from the frontend to the api-route so that we can query the api with that value.

PS: The API expects a parameter like `barking=1` or `name=dachshund`, and so on ...

TY

**Comment by NikRoe** [comment]
Hi there,

you should be able to send this information to your own api and read that information before sending the request to the external API.

When you fetch your data, you currently have this code:

```js
const { data } = useSWR("/api/quizResults", fetcher);
```

You should be able to extend the URL to also contain the parameters and look like this:

```js
const { data } = useSWR(`/api/quizResults?barking={hereYouShouldEnterTheBarkingValueFromTheUser}&energy={energyValueFromTheUser}`, fetcher);
```

You should be able to extract that information in your api route from the request object.
Edit: in the `request.query`, to be more precise

Hope this already helps. Let me know if you need further assistance ✌️

**Reply by moonwave99** [reply, marked as answer]
That's it ^^

To be future proof, one can use [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) like:

```js
const params = new URLSearchParams({
  barking: valueFromSomeWhere,
  energy: ...,
  ...
});

useSWR(`/api/quizResults?${params}`);
```

> We have three questions so far, regarding barking

Gotta love that!

**Comment by alinarub** [comment]
Thanks so much for both of your answers! That looks like a great solution for the frontend. How does the backend receive the search parameters? This is what our backend currently looks like:

```js
export default async function handler(request, response) {

  if (request.method === "GET") {
    // make request to API. then send back response.
    // this example does not use any API KEYS.
    // If you need to keep something secret you could do the following. {process.env.YOUR_NAME_OF_THE_ENV_VARIABLE}

    const dogResponse = await fetch(
      `https://api.api-ninjas.com/v1/dogs?barking=5`,
      {
        headers: {
          "X-Api-Key": process.env.DOG_API_KEY,
        },
      }
    );
    const data = await dogResponse.json();
 
    response.status(200).json(data);
    return;
  }

  response.status(405).json({ message: "Method not allowed" });
}
```

**Reply by NikRoe** [reply]
The information is stored in the request object.

```js
export default async function handler(request, response) {

const queryParameters = request.query;
// this will be an object with the query parameters
// working with a console.log() here might be beneficial to see how the data looks like

if (request.method === "GET") {
...
}
```

**Reply by alinarub** [reply]
We saw your answer after posting 😀

**Reply by moonwave99** [reply]
...and the whole thing will look like:

```js
const params = new URLSearchParams(request.query);
const dogResponse = await fetch(
  `https://api.api-ninjas.com/v1/dogs?${params}`,
  {
    headers: {
      "X-Api-Key": process.env.DOG_API_KEY,
    },
  }
);
```

Or if you want to have some defaults:

```js
const params = new URLSearchParams({
  barking: x,
  trainability: y,
  energy: z,
  ...request.query
});
const dogResponse = await fetch(
  `https://api.api-ninjas.com/v1/dogs?${params}`,
  {
    headers: {
      "X-Api-Key": process.env.DOG_API_KEY,
    },
  }
);
```

(sorry @NikRoe for the intermission 🐶)

**Reply by NikRoe** [reply]
all good, all good 👌

**Comment by alinarub** [comment]
We managed to get a list of all dogs with barking: 2 in the frontend with the help of request.query.barking. 🎉 

Apart from barking, we also have energy and trainability. Our API can't chain multiple parameters at the same time, that's why we decided to fetch three times for three different parameters. How can we make the parameter and the value dynamic, so that we don't have to hardcode the rating number and the category? E.g. instead of barking=5, we can also receive energy=3.

This is what we currently have: 

      https://api.api-ninjas.com/v1/dogs?barking=${request.query.barking}
     

Thanks again!

**Reply by NikRoe** [reply]
I am not entirely sure about the second paragraph. 

Let's meet tomorrow morning and talk about it in person 👌
